<a href="https://colab.research.google.com/github/HydrOpOtesJA99/A.I-learning/blob/main/13_%ED%9E%88%EB%93%A0%EB%A0%88%EC%9D%B4%EC%96%B4_%EC%8B%A4%EC%8A%B5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 13_히든레이어_실습

*   보스턴 집값 예측: https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv
*   아이리스 품종 분류: https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv



In [4]:
# 라이브러리 사용

import tensorflow as tf
import pandas as pd

## 1. 보스턴 집값 예측

### 1) 과거의 데이터 준비

In [34]:
# 과거의 데이터 준비
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
보스턴 = pd.read_csv(파일경로)

# 보스턴.head() # 파일경로 제대로 연결되었는지 확인
print(보스턴.columns)

# 종속변수, 독립변수 준비
독립 = 보스턴[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 
              'age', 'dis', 'rad', 'tax', 'ptratio', 'b', 'lstat']]
종속 = 보스턴[['medv']]

print('독립변수(행,열):',독립.shape, ', 종속변수(행,열):',종속.shape)

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')
독립변수(행,열): (506, 13) , 종속변수(행,열): (506, 1)


### 2) 모델의 구조를 만든다

In [36]:
# 모델의 구조를 만든다
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(10, activation ='swish')(X)
Y = tf.keras.layers.Dense(1)(H) # 'H'로 바꾸는것 잊지말기
model = tf.keras.models.Model(X,Y)
model.compile(loss='mse')

### 2-1) 만든 모델의 확인

In [43]:
# 만든 모델의 확인
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, 13)]              0         
                                                                 
 dense (Dense)               (None, 10)                140       
                                                                 
 dense_1 (Dense)             (None, 1)                 11        
                                                                 
Total params: 151
Trainable params: 151
Non-trainable params: 0
_________________________________________________________________




*   param(parameter; 가중치) 
*   이전 Output Shape 값에 +1(Bias)을 하고 다음 Output Shape 곱하면 Param 값이 나온다
*   ex)(13(output shape)+1(bias)) * 10(output shape) = 140(Param)




### 3) 모델을 학습시킨다.

In [47]:
# 모델을 학습시킨다
model.fit(독립, 종속, epochs=1000, verbose=0)
model.fit(독립, 종속, epochs=10)

Epoch 1/10
16/16 [==============================] - 0s 2ms/step - loss: 11.9324
Epoch 2/10
16/16 [==============================] - 0s 2ms/step - loss: 10.9011
Epoch 3/10
16/16 [==============================] - 0s 2ms/step - loss: 11.1768
Epoch 4/10
16/16 [==============================] - 0s 2ms/step - loss: 10.8364
Epoch 5/10
16/16 [==============================] - 0s 2ms/step - loss: 11.4163
Epoch 6/10
16/16 [==============================] - 0s 2ms/step - loss: 10.2545
Epoch 7/10
16/16 [==============================] - 0s 2ms/step - loss: 10.9989
Epoch 8/10
16/16 [==============================] - 0s 2ms/step - loss: 11.2385
Epoch 9/10
16/16 [==============================] - 0s 2ms/step - loss: 10.1764
Epoch 10/10
16/16 [==============================] - 0s 2ms/step - loss: 11.4016


### 4) 모델을 이용한다

In [48]:
# 모델을 이용한다

print(model.predict(독립[:5]))
print(종속[:5])

[[30.264984]
 [22.23278 ]
 [33.96098 ]
 [33.34128 ]
 [29.68676 ]]
   medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2


## 2. 아이리스 품종 분류

### 1) 과거의 데이터 준비

In [55]:
# 과거의 데이터 준비

파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
아이리스 = pd.read_csv(파일경로)
아이리스.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

### 1-1) 원핫인코딩(범주형 데이터)

In [57]:
# 원핫인코딩

아이리스 = pd.get_dummies(아이리스)
아이리스.head
#print(아이리스.columns)

<bound method NDFrame.head of      꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0     5.1  3.5    1.4   0.2          1              0             0
1     4.9  3.0    1.4   0.2          1              0             0
2     4.7  3.2    1.3   0.2          1              0             0
3     4.6  3.1    1.5   0.2          1              0             0
4     5.0  3.6    1.4   0.2          1              0             0
..    ...  ...    ...   ...        ...            ...           ...
145   6.7  3.0    5.2   2.3          0              0             1
146   6.3  2.5    5.0   1.9          0              0             1
147   6.5  3.0    5.2   2.0          0              0             1
148   6.2  3.4    5.4   2.3          0              0             1
149   5.9  3.0    5.1   1.8          0              0             1

[150 rows x 7 columns]>

In [59]:
# 종속변수, 독립변수 설정

독립 = 아이리스[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속 = 아이리스[['품종_setosa', '품종_versicolor', '품종_virginica']]
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


### 2) 모델의 구조를 만든다

In [62]:
# 모델의 구조를 만든다

X = tf.keras.layers.Input(shape=[4])
H = tf.keras.layers.Dense(8, activation = 'swish')(X)
H = tf.keras.layers.Dense(8, activation = 'swish')(X)
H = tf.keras.layers.Dense(8, activation = 'swish')(X)
Y = tf.keras.layers.Dense(3, activation = 'softmax')(H)
model = tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy', metrics='accuracy')

### 2-1) 만든 모델의 확인

In [65]:
# 만든 모델의 확인

model.summary()

Model: "model_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_5 (InputLayer)        [(None, 4)]               0         
                                                                 
 dense_6 (Dense)             (None, 8)                 40        
                                                                 
 dense_7 (Dense)             (None, 3)                 27        
                                                                 
Total params: 67
Trainable params: 67
Non-trainable params: 0
_________________________________________________________________


### 3) 모델을 학습시킨다.

In [63]:
# 모델을 학습시킨다

model.fit(독립, 종속, epochs=1000, verbose=0)
model.fit(독립, 종속, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 5ms/step - loss: 0.0615 - accuracy: 0.9867
Epoch 2/10
5/5 [==============================] - 0s 5ms/step - loss: 0.0627 - accuracy: 0.9667
Epoch 3/10
5/5 [==============================] - 0s 5ms/step - loss: 0.0601 - accuracy: 0.9867
Epoch 4/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0616 - accuracy: 0.9733
Epoch 5/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0603 - accuracy: 0.9800
Epoch 6/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0605 - accuracy: 0.9667
Epoch 7/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0604 - accuracy: 0.9733
Epoch 8/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0605 - accuracy: 0.9800
Epoch 9/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0612 - accuracy: 0.9600
Epoch 10/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0603 - accuracy: 0.9667


### 4) 모델을 이용한다

In [64]:
print(model.predict(독립[0:5]))
print(종속[0:5])

[[9.9973017e-01 2.5415749e-04 1.5623278e-05]
 [9.9974579e-01 2.5047749e-04 3.7267823e-06]
 [9.9968112e-01 3.0137127e-04 1.7468623e-05]
 [9.9970180e-01 2.9484762e-04 3.3615572e-06]
 [9.9968827e-01 2.8665442e-04 2.5126754e-05]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0




*   첫번째 행에서 
*   품종_setosa일 확률이 9.99..e-01(9.99...^-1)
*   품종_versicolor일 확률이 2.54...e-04
*   품종_virginica일 확률이 1.56...e-05

